share/man/man4/joy.4
.\" $NetBSD$
.\"
.\" Copyright (c) 1996 Matthieu Herrb
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed by Christopher G. Demetriou.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd July 22, 2006
.Dt JOY 4
.Os
.Sh NAME
.Nm joy
.Nd game adapter driver
.Sh SYNOPSIS
.Cd "joy* at acpi?"
.Cd "joy* at eap?"
.Cd "joy* at eso?"
.Cd "joy0 at isa? port 0x201"
.Cd "joy* at isapnp?"
.Cd "joy* at ofisa?"
.Cd "joy* at pci?"
.Cd "joy* at pnpbios? index ?"
.Sh DESCRIPTION
This driver provides access to the game adapter.
The lower bit in the minor device number selects the joystick: 0
is the first joystick and 1 is the second.
.Pp
The game control adapter allows up to two joysticks to be attached to
the system.
The adapter plus the driver convert the present resistive value to
a relative joystick position.
On receipt of an output signal, four timing circuits are started.
By determining the time required for the circuit to time-out (a
function of the resistance), the paddle position can be determined.
The adapter could be used as a general purpose I/O card with four
analog (resistive) inputs plus four digital input points.
.Pp
Applications may call
.Xr ioctl 2
on a game adapter driver file descriptor
to set and get the offsets of the two potentiometers and the maximum
time-out value for the circuit.
The
.Xr ioctl 2
commands are listed in
.In machine/joystick.h
and currently are:
.Pp
.Bl -tag -width JOY_GET_X_OFFSET -compact
.It Dv JOY_SETTIMEOUT
Sets the maximum time-out for the adapter.
.It Dv JOY_GETTIMEOUT
Returns the current maximum time-out.
.It Dv JOY_SET_X_OFFSET
Sets an offset on X value.
.It Dv JOY_GET_X_OFFSET
Returns the current X offset.
.It Dv JOY_SET_Y_OFFSET
Sets an offset on Y value.
.It Dv JOY_GET_Y_OFFSET
Returns the current Y offset.
.El
.Pp
All these commands take an integer parameter.
.Pp
.Xr read 2
on the file descriptor returns a
.Fa joystick
structure:
.Bd -literal -offset indent
struct joystick {
	int x;
	int y;
	int b1;
	int b2;
};
.Ed
.Pp
The fields have the following functions:
.Bl -tag -width xxx
.It Fa x
current X coordinate of the joystick (or position of paddle 1)
.It Fa y
current Y coordinate of the joystick (or position of paddle 2)
.It Fa b1
current state of button 1
.It Fa b2
current state of button 2
.El
.Pp
The b1 and b2 fields in struct joystick are set to 1 if the
corresponding button is down, 0 otherwise.
.Pp
The x and y coordinates are supposed to be between 0 and 255 for a
good joystick and a good adapter.
Unfortunately, because of the
hardware hack that is used to measure the position (by measuring the
time needed to discharge an RC circuit made from the joystick's
potentiometer and a capacitor on the adapter), calibration
is needed to determine exactly what values are returned for a specific
joystick/adapter combination.
Incorrect hardware can yield negative or values greater than 255.
.Pp
A typical calibration procedure uses the values returned at lower
left, center and upper right positions of the joystick to compute the
relative position.
.Pp
This calibration is not part of the driver.
.Sh FILES
.Bl -tag -width Pa -compact
.It Pa /dev/joy0
first joystick
.It Pa /dev/joy1
second joystick
.El
.Sh SEE ALSO
.Xr acpi 4 ,
.Xr eap 4 ,
.Xr eso 4 ,
.Xr isa 4 ,
.Xr isapnp 4 ,
.Xr ofisa 4 ,
.Xr pci 4 ,
.Xr pnpbios 4
.Sh AUTHORS
Jean-Marc Zucconi wrote the
.Fx
driver.
Matthieu Herrb ported it to
.Nx
and wrote this manual page.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

sys/dev/ic/joy.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 2008 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software developed for The NetBSD Foundation
 * by Andrew Doran.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*-
 * Copyright (c) 1995 Jean-Marc Zucconi
 * All rights reserved.
 *
 * Ported to NetBSD by Matthieu Herrb <matthieu@laas.fr>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer
 *    in this position and unchanged.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/kernel.h>
#include <sys/device.h>
#include <sys/errno.h>
#include <sys/conf.h>
#include <sys/event.h>
#include <sys/vnode.h>
#include <sys/bus.h>
#include <sys/joystick.h>

#include <dev/ic/joyvar.h>

/*
 * The game port can manage 4 buttons and 4 variable resistors (usually 2
 * joysticks, each with 2 buttons and 2 pots.) via the port at address 0x201.
 * Getting the state of the buttons is done by reading the game port;
 * buttons 1-4 correspond to bits 4-7 and resistors 1-4 (X1, Y1, X2, Y2)
 * to bits 0-3.  If button 1 (resp 2, 3, 4) is pressed, the bit 4 (resp 5,
 * 6, 7) is set to 0 to get the value of a resistor, write the value 0xff
 * at port and wait until the corresponding bit returns to 0.
 */


#define JOYPART(d) (minor(d) & 1)
#define JOYUNIT(d) (minor(d) >> 1)

#ifndef JOY_TIMEOUT
#define JOY_TIMEOUT   2000	/* 2 milliseconds */
#endif

extern struct cfdriver joy_cd;

static dev_type_open(joyopen);
static dev_type_close(joyclose);
static dev_type_read(joyread);
static dev_type_ioctl(joyioctl);

const struct cdevsw joy_cdevsw = {
	joyopen, joyclose, joyread, nowrite, joyioctl,
	nostop, notty, nopoll, nommap, nokqfilter, D_OTHER | D_MPSAFE,
};

void
joyattach(struct joy_softc *sc)
{

	if (sc->sc_lock == NULL) {
		panic("joyattach: no lock");
	}

	sc->timeout[0] = 0;
	sc->timeout[1] = 0;

	mutex_enter(sc->sc_lock);
	bus_space_write_1(sc->sc_iot, sc->sc_ioh, 0, 0xff);
	DELAY(10000);		/* 10 ms delay */
	aprint_normal_dev(sc->sc_dev, "joystick %sconnected\n",
	    (bus_space_read_1(sc->sc_iot, sc->sc_ioh, 0) & 0x0f) == 0x0f ?
	    "not " : "");
	mutex_exit(sc->sc_lock);
}

int
joydetach(struct joy_softc *sc, int flags)
{
	int maj, mn;

	maj = cdevsw_lookup_major(&joy_cdevsw);
	mn = device_unit(sc->sc_dev) << 1;
	vdevgone(maj, mn, mn, VCHR);
	vdevgone(maj, mn + 1, mn + 1, VCHR);

	return 0;
}

static int
joyopen(dev_t dev, int flag, int mode, struct lwp *l)
{
	int unit = JOYUNIT(dev);
	int i = JOYPART(dev);
	struct joy_softc *sc;

	sc = device_lookup_private(&joy_cd, unit);
	if (sc == NULL)
		return ENXIO;

	mutex_enter(sc->sc_lock);
	if (sc->timeout[i]) {
		mutex_exit(sc->sc_lock);
		return EBUSY;
	}
	sc->x_off[i] = sc->y_off[i] = 0;
	sc->timeout[i] = JOY_TIMEOUT;
	mutex_exit(sc->sc_lock);
	return 0;
}

static int
joyclose(dev_t dev, int flag, int mode, struct lwp *l)
{
	int unit = JOYUNIT(dev);
	int i = JOYPART(dev);
	struct joy_softc *sc = device_lookup_private(&joy_cd, unit);

	mutex_enter(sc->sc_lock);
	sc->timeout[i] = 0;
	mutex_exit(sc->sc_lock);
	return 0;
}

static int
joyread(dev_t dev, struct uio *uio, int flag)
{
	int unit = JOYUNIT(dev);
	struct joy_softc *sc = device_lookup_private(&joy_cd, unit);
	bus_space_tag_t iot = sc->sc_iot;
	bus_space_handle_t ioh = sc->sc_ioh;
	struct joystick c;
	struct timeval start, now, diff;
	int state = 0, x = 0, y = 0, i;

	mutex_enter(sc->sc_lock);
	bus_space_write_1(iot, ioh, 0, 0xff);
	microtime(&start);
	now = start; /* structure assignment */
	i = sc->timeout[JOYPART(dev)];
	for (;;) {
		timersub(&now, &start, &diff);
		if (diff.tv_sec > 0 || diff.tv_usec > i)
			break;
		state = bus_space_read_1(iot, ioh, 0);
		if (JOYPART(dev) == 1)
			state >>= 2;
		if (!x && !(state & 0x01))
			x = diff.tv_usec;
		if (!y && !(state & 0x02))
			y = diff.tv_usec;
		if (x && y)
			break;
		microtime(&now);
	}
	mutex_exit(sc->sc_lock);

	c.x = x ? sc->x_off[JOYPART(dev)] + x : 0x80000000;
	c.y = y ? sc->y_off[JOYPART(dev)] + y : 0x80000000;
	state >>= 4;
	c.b1 = ~state & 1;
	c.b2 = ~(state >> 1) & 1;
	return uiomove(&c, sizeof(struct joystick), uio);
}

static int
joyioctl(dev_t dev, u_long cmd, void *data, int flag, struct lwp *l)
{
	int unit = JOYUNIT(dev);
	struct joy_softc *sc = device_lookup_private(&joy_cd, unit);
	int i = JOYPART(dev), x, error;

	mutex_enter(sc->sc_lock);
	error = 0;
	switch (cmd) {
	case JOY_SETTIMEOUT:
		x = *(int *)data;
		if (x < 1 || x > 10000) {	/* 10ms maximum! */
			error = EINVAL;
			break;
		}
		sc->timeout[i] = x;
		break;
	case JOY_GETTIMEOUT:
		*(int *)data = sc->timeout[i];
		break;
	case JOY_SET_X_OFFSET:
		sc->x_off[i] = *(int *)data;
		break;
	case JOY_SET_Y_OFFSET:
		sc->y_off[i] = *(int *)data;
		break;
	case JOY_GET_X_OFFSET:
		*(int *)data = sc->x_off[i];
		break;
	case JOY_GET_Y_OFFSET:
		*(int *)data = sc->y_off[i];
		break;
	default:
		error = ENXIO;
		break;
	}
	mutex_exit(sc->sc_lock);
	return error;
}
